omf-python/src/attribute.rs

```diff
@@ -0,0 +1,437 @@
+    /// Selects which part of the element the attribute is attached to.
+    ///
+    /// See the documentation for each Geometry variant for a list of what locations are valid.
+    fn location(&self) -> PyLocation {
+        match self.0.location {
+            Location::Vertices => PyLocation::Vertices,
+            Location::Primitives => PyLocation::Primitives,
+            Location::Subblocks => PyLocation::Subblocks,
+            Location::Elements => PyLocation::Elements,
+            Location::Projected => PyLocation::Projected,
+            Location::Categories => PyLocation::Categories,
+        }
+    }
+
+    #[getter]
+    /// The attribute data as a JSON string.
+    fn data_json(&self) -> PyResult<String> {
+        let data = serde_json::to_string(&self.0.data)
+            .map_err(|e| OmfException::py_err(DeserializationFailed(e)))?;
+        Ok(data)
+    }
+
+    /// The attribute data.
+    fn get_data(&self, py: Python<'_>) -> PyResult<PyObject> {
+        match &self.0.data {
+            AttributeData::Category { .. } => {
+                Ok(PyAttributeDataCategory(self.0.data.clone()).into_py(py))
+            }
+            AttributeData::Color { .. } => {
+                Ok(PyAttributeDataColor(self.0.data.clone()).into_py(py))
+            }
+            AttributeData::MappedTexture { .. } => {
+                Ok(PyAttributeDataMappedTexture(self.0.data.clone()).into_py(py))
+            }
+            AttributeData::ProjectedTexture { .. } => {
+                Ok(PyAttributeDataProjectedTexture(self.0.data.clone()).into_py(py))
+            }
+            AttributeData::Number { .. } => {
+                Ok(PyAttributeDataNumber(self.0.data.clone()).into_py(py))
+            }
+            AttributeData::Vector { .. } => {
+                Ok(PyAttributeDataVector(self.0.data.clone()).into_py(py))
+            }
+            AttributeData::Boolean { .. } => {
+                Ok(PyAttributeDataBoolean(self.0.data.clone()).into_py(py))
+            }
+            AttributeData::Text { .. } => Ok(PyAttributeDataText(self.0.data.clone()).into_py(py)),
+        }
+    }
+}
+
+#[gen_stub_pyclass]
+#[pyclass(name = "AttributeDataCategory")]
+/// Category data.
+///
+/// A name is required for each category, a color is optional, and other values can be attached
+/// as sub-attributes.
+pub struct PyAttributeDataCategory(AttributeData);
+
+#[gen_stub_pymethods]
+#[pymethods]
+impl PyAttributeDataCategory {
+    #[getter]
+    /// Array with `Index` type storing the category indices.
+    ///
+    /// Values are indices into the `names` array, `colors` array, and any sub-attributes,
+    /// and must be within range for them.
+    fn values(&self) -> PyResult<PyIndexArray> {
+        match &self.0 {
+            AttributeData::Category { values, .. } => Ok(PyIndexArray(values.clone())),
+            _ => Err(OmfNotSupportedException::new_err(
+                "AttributeData variant is not supported",
+            )),
```

Comment:
Again, I think it's tidier for this to panic.

Also, this pattern is repeated a lot here. Using a macro would make it more readable. This function could be reduced to:
```rust
fn values(&self) -> PyIndexArray {
    attribute_data_field!(self, Category::values).clone()
}
```
With a macro like this (untested):
```rust
macro_rules! attribute_data_field {
    ($self:ident, $variant:ident :: $name:ident) => {
        match &$self.0 {
            AttributeData::$variant { $field, .. } => &$field,
            _ => unreachable!(),
        }
    };
}
```